auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/external/DeltaUserRegistrationController.kt

```diff
@@ -0,0 +1,241 @@
+        if (hasDisallowedSpecialCharacters(emailAddress)) emailAddressErrors.add(emailContainsDisallowedCharacters)
+
+        var organisations: List<Organisation> = listOf()
+        if (emailAddress.isEmpty()) emailAddressErrors.add(emailAddressEmpty)
+        else {
+            if (!emailAddressChecker.hasValidEmailFormat(emailAddress)) emailAddressErrors.add(notAnEmailAddress)
+            else {
+                organisations = organisationService.findAllByDomain(emailToDomain(emailAddress))
+                if (!emailAddressChecker.hasKnownNotRetiredDomain(emailAddress, organisations)) emailAddressErrors.add(
+                    notAKnownDomain
+                )
+            }
+        }
+        if (confirmEmailAddress.isEmpty()) confirmEmailAddressErrors.add(confirmEmailAddressEmpty)
+        else if (confirmEmailAddress != emailAddress) confirmEmailAddressErrors.add(notEqualEmails)
+
+        if (hasErrors(firstNameErrors, lastNameErrors, emailAddressErrors, confirmEmailAddressErrors)) {
+            return call.respondRegisterPage(
+                firstName,
+                lastName,
+                emailAddress,
+                confirmEmailAddress,
+                firstNameErrors,
+                lastNameErrors,
+                emailAddressErrors,
+                confirmEmailAddressErrors
+            )
+        } else {
+            val ssoClientMatchingEmailDomain = ssoConfig.ssoClients.firstOrNull {
+                it.required && emailAddress.lowercase().endsWith(it.emailDomain)
+            }
+            if (ssoClientMatchingEmailDomain != null) {
+                return call.respondRedirect(
+                    deltaRouteWithEmail(
+                        deltaConfig.deltaWebsiteUrl,
+                        ssoClientMatchingEmailDomain.internalId,
+                        emailAddress
+                    )
+                )
+            }
+
+            val registration = Registration(firstName, lastName, emailAddress)
+            lateinit var registrationResult: RegistrationService.RegistrationResult
+            try {
+                registrationResult = registrationService.register(registration, organisations)
+            } catch (e: Exception) {
+                logger.error(
+                    "Error registering user with  name: {} {}, email address: {}",
+                    firstName,
+                    lastName,
+                    emailAddress,
+                    e
+                )
+                throw e // TODO - handle errors in a tidier way so these two can be combined - make specific exceptions
+            }
+            try {
+                registrationService.sendRegistrationEmail(registrationResult)
+            } catch (e: Exception) {
+                // If it fails on this step the user never gets a link and doesn't know if account is made - TODO - is there a solution to this?
```

Comment:
Not really 🤷 we could queue emails and retry them, but not worth the trouble with the small amount we send